File: internal/db/flush_test.go
package db

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestBackgroundFlushDoesNotBreakReads(t *testing.T) {
	dir := t.TempDir()

	opts := DefaultOptions()
	opts.DataDir = dir
	opts.WALFileName = "wal.log"
	opts.MemtableThreshold = 2
	opts.SparseIndexGap = 1

	database, err := Open(opts)
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	defer database.Close()

	if err := database.Put([]byte("a"), []byte("1")); err != nil {
		t.Fatalf("put a: %v", err)
	}
	if err := database.Put([]byte("b"), []byte("2")); err != nil {
		t.Fatalf("put b: %v", err)
	}
	if err := database.Put([]byte("c"), []byte("3")); err != nil {
		t.Fatalf("put c: %v", err)
	}

	value, ok, err := database.Get([]byte("a"))
	if err != nil {
		t.Fatalf("get a: %v", err)
	}
	if !ok || string(value) != "1" {
		t.Fatalf("unexpected read for a: ok=%v value=%q", ok, value)
	}

	sstDir := filepath.Join(dir, opts.SSTableDirName)
	found := false
	deadline := time.Now().Add(2 * time.Second)

	for time.Now().Before(deadline) {
		entries, err := os.ReadDir(sstDir)
		if err != nil {
			t.Fatalf("read sst dir: %v", err)
		}
		if len(entries) > 0 {
			found = true
			break
		}
		time.Sleep(20 * time.Millisecond)
	}

	if !found {
		t.Fatal("expected background flush to create an sstable")
	}
}

File: internal/db/flush_test.cpp
#include "db.h"
#include <gtest/gtest.h>
#include <filesystem>
#include <thread>
#include <chrono>

using namespace bigdb::db;
namespace fs = std::filesystem;

class FlushTest : public ::testing::Test {
protected:
    void SetUp() override {
        temp_dir = testing::TempDir() + "/flush_test_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
        fs::create_directories(temp_dir);
    }
    void TearDown() override {
        fs::remove_all(temp_dir);
    }
    std::string temp_dir;
};

TEST_F(FlushTest, BackgroundFlushDoesNotBreakReads) {
    Options opts = DefaultOptions();
    opts.DataDir = temp_dir;
    opts.WALFileName = "wal.log";
    opts.MemtableThreshold = 2;
    opts.SparseIndexGap = 1;

    auto db = DB::Open(opts);
    ASSERT_NE(db, nullptr);

    EXPECT_NO_THROW(db->Put({'a'}, {'1'}));
    EXPECT_NO_THROW(db->Put({'b'}, {'2'}));
    EXPECT_NO_THROW(db->Put({'c'}, {'3'}));

    auto [value, ok] = db->Get({'a'});
    ASSERT_TRUE(ok);
    std::string valA(value.begin(), value.end());
    EXPECT_EQ(valA, "1");

    std::string sstDir = (fs::path(temp_dir) / opts.SSTableDirName).string();
    bool found = false;
    
    auto deadline = std::chrono::system_clock::now() + std::chrono::seconds(2);
    while (std::chrono::system_clock::now() < deadline) {
        if (fs::exists(sstDir)) {
            for (const auto& entry : fs::directory_iterator(sstDir)) {
                if (entry.is_regular_file()) {
                    found = true;
                    break;
                }
            }
        }
        if (found) break;
        std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }

    EXPECT_TRUE(found) << "expected background flush to create an sstable";
    db->Close();
}
